keps/sig-api-machinery/4222-cbor-serializer/README.md

```diff
@@ -0,0 +1,1209 @@
+Sequence](https://www.rfc-editor.org/rfc/rfc8742.html) is a concatenation of
+zero or more CBOR data items, with no additional framing. This is effectively
+equivalent to the existing JSON stream serialization behavior and takes
+advantage of the property that, like JSON objects – and unlike non-object JSON
+documents, e.g. numbers – CBOR data items are self-delimiting.
+
+### Negotiation
+
+Proactive content negotiation will be supported for clients that want to receive
+CBOR-encoded responses using the MIME type “application/cbor” in the Accept
+request header. For compatibility with API servers that don’t support CBOR,
+clients should also accept “application/json” (with a lower quality factor) and
+choose the appropriate decoder based on the Content-Type response header.
+
+Similarly, streaming requests should use the MIME type for CBOR Sequences,
+“application/cbor-seq”.
+
+CBOR will not be a supported encoding for JSON Patches, JSON Merge Patches,
+Strategic Merge Patches, or Server-Side Apply configurations.
+
+Clients can send CBOR-encoded request bodies with the appropriate Content-Type
+to API servers that support CBOR. API servers that don’t support CBOR will
+return status 415 (Unsupported Media Type). In client-go, for alpha, when a
+RESTClient configured to encode requests with CBOR receives a 415, it will
+permanently (for the life of the RESTClient) fall back to JSON for subsequent
+requests. For GA, this fallback behavior will be changed to operate on a
+per-(method, target resource) basis, and to consider acceptable fallback
+content-types based on the value of the Accept header in a 415 response, [as
+described in RFC 9110](https://httpwg.org/specs/rfc9110.html#status.415).
+
+### Client Enablement
+
+Clients can be explicitly configured to prefer CBOR as a request encoding as
+they can today be configured to prefer Protobuf or JSON. In client-go, this
+involves setting the `ContentType` field of `rest.ClientContentConfig`. The
+default request content-type will remain JSON for a period of time post-GA; a
+minimum of two minor versions, so that the oldest kube-apiserver within the
+supported kubectl version skew will have CBOR support. The supported version
+skew for aggregated API servers is much wider (infinite?). Encoding and decoding
+resources from aggregated API servers that don't support CBOR will rely on the
+content-type negotation mechanisms described above.
+
+CBOR enablement will be feature-gated in API servers. Client-go will provide a
+runtime gating mechanism for CBOR request encoding as follows:
+
+- An environment variable will be read at package initialization time with
+  recognized values:
+  - "default": The default request content-type (if not explicitly configured)
+    becomes application/cbor rather than application/json.
+  - "disabled": Clients configured with request content-type application/cbor
+    behave as if configured to use application/json.
```

Comment:
Notably, these options don't allow the consumer of a binary to force-enable writes in CBOR if the author of the binary has explicitly configured the client with a different content-type.